common/recordCreate.js

```diff
@@ -0,0 +1,300 @@
+         *
+         * @param  {bool} isUpdate            flag
+         * @param  {object} derivedref        derived referene of the column that is being added
+         * @param  {array} recordEditModel   data array
+         * @param  {bool} isModalUpdate     if updating through record app
+         * @param  {object} onSuccessFunction callback
+         */
+        function addRecords(isUpdate, derivedref, recordEditModel, isModalUpdate, onSuccessFunction) {
+            var model = isModalUpdate ? GV_recordEditModel : recordEditModel;
+            var form = viewModel.formContainer;
+
+            // this will include updated and previous raw values.
+            var submissionRowsCopy = [];
+
+            model.submissionRows.forEach(function(row) {
+                submissionRowsCopy.push(Object.assign({}, row));
+            });
+
+            /**
+             * Add raw values that are not visible to submissionRowsCopy:
+             *
+             * submissionRowsCopy is the datastructure that will be used for creating
+             * the upload url. It must have all the visible and invisible data.
+             * The following makes sure that submissionRowsCopy has all the underlying data
+             */
+            if (isUpdate) {
+                for (var i = 0; i < submissionRowsCopy.length; i++) {
+                    var newData = submissionRowsCopy[i];
+                    var oldData = $rootScope.tuples[i].data;
+
+                    // make sure submissionRowsCopy has all the data
+                    for (var key in oldData) {
+                        if (key in newData) continue;
+                        newData[key] = oldData[key];
+                    }
+                }
+            }
+
+            //call uploadFiles which will upload files and callback on success
+            uploadFiles(submissionRowsCopy, isUpdate, function() {
+
+                var fn = "create",
+                    args = [submissionRowsCopy];
+                var fnScope = isModalUpdate ? derivedref.unfilteredReference.contextualize.entryCreate : $rootScope.reference.unfilteredReference.contextualize.entryCreate;
+
+                if (isUpdate) {
+
+                    /**
+                     * After uploading files, the returned submissionRowsCopy contains
+                     * new file data. This includes filename, filebyte, and md5.
+                     * The following makes sure that all the data are updated.
+                     * That's why this for loop must be after uploading files and not before.
+                     * And we cannot just pass submissionRowsCopy to update function, because
+                     * update function only accepts array of tuples (and not just key-value pair).
+                     */
+                    for (var i = 0; i < submissionRowsCopy.length; i++) {
+                        var row = submissionRowsCopy[i];
+                        var data = $rootScope.tuples[i].data;
```

Comment:
Same comment as above about passing `tuples` in as a parameter and making sure it's not on the model beforehand.